internal/users/db/migration.go

```diff
@@ -0,0 +1,316 @@
+	log.Debugf(context.Background(), "Renaming users in %q: %v -> %v", groupFile,
+		oldNames, newNames)
+
+	// Note that we can't use gpasswd here because `gpasswd --add` checks for the existence of the user, which causes an
+	// NSS request to be sent to authd, but authd is not ready yet because we are still migrating the database.
+	err := userutils.WriteLockShadowPassword()
+	if err != nil {
+		return fmt.Errorf("failed to lock group file: %w", err)
+	}
+	defer func() {
+		if err := userutils.WriteUnlockShadowPassword(); err != nil {
+			log.Warningf(context.Background(), "Failed to unlock group file: %v", err)
+		}
+	}()
+
+	content, err := os.ReadFile(groupFile)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", groupFile, err)
+	}
+
+	oldLines := strings.Split(string(content), "\n")
+	var newLines []string
+
+	for _, line := range oldLines {
+		if line == "" {
+			continue
+		}
+
+		fields := strings.FieldsFunc(line, func(r rune) bool { return r == ':' })
+		if len(fields) != 4 {
+			return fmt.Errorf("unexpected number of fields in %s line (expected 4, got %d): %s",
+				groupFile, len(fields), line)
+		}
+
+		users := strings.Split(fields[3], ",")
+		for j, user := range users {
+			for k, oldName := range oldNames {
+				if user == oldName {
+					users[j] = newNames[k]
+				}
+			}
+		}
+
+		fields[3] = strings.Join(users, ",")
+		newLines = append(newLines, strings.Join(fields, ":"))
+	}
+
+	// Add final new line to the group file.
+	newLines = append(newLines, "")
+
+	backupPath := groupFileBackupPath()
+	backupDone := true
+	oldBackup := ""
+
+	if tmpDir, err := os.MkdirTemp(os.TempDir(), "authd-migration-backup"); err == nil {
+		b := filepath.Join(tmpDir, filepath.Base(backupPath))
+		err := fileutils.CopyFile(backupPath, b)
+		if err == nil {
+			oldBackup = b
+		}
+		if err != nil && !errors.Is(err, os.ErrNotExist) {
+			log.Warningf(context.Background(), "Failed to create backup of %q: %v", backupPath, err)
+		}
+		defer os.Remove(oldBackup)
+	}
+
+	if err := os.Remove(backupPath); err != nil && !errors.Is(err, os.ErrNotExist) {
+		log.Warningf(context.Background(), "Failed to remove group file backup: %v", err)
+	}
+	if err := os.Rename(groupFile, backupPath); err != nil {
+		log.Warningf(context.Background(), "Failed make a backup for the group file: %v", err)
+
+		if oldBackup != "" {
+			// Backup of current group file failed, let's restore the old backup.
+			_ = os.Rename(oldBackup, backupPath)
+		}
+		backupDone = false
+	}
+
+	//nolint:gosec // G306 /etc/group should indeed have 0644 permissions
+	if err := os.WriteFile(groupFile, []byte(strings.Join(newLines, "\n")), 0644); err != nil {
+		retErr := fmt.Errorf("error writing %s: %w", groupFile, err)
+		if !backupDone {
+			return retErr
+		}
+
+		if err := os.Rename(backupPath, groupFile); err != nil {
+			log.Errorf(context.Background(), "Failed to restore the backup for the group file: %v", err)
+		}
```

Comment:
We can omit this when we rename a temporary file. I also don't see gpasswd ever restoring or deleting the backup file it creates, AFAICT it just keeps it around as `/etc/group-`.

Reply:
> AFAICT it just keeps it around as `/etc/group-`

I suspect that's done so that the user can manually restore the backup in case `gpasswd` was called accidentally. I don't think we need that, because the changes done by authd are done without user interaction. So IMO we don't need to create a backup file at all.

Reply:
Well, I feel backups are always better to exist than the other way around... Since the code is now there (without adding much complexity) and tested (and won't cause troubles in case of backup failures), IMHO is still better to keep a backup (named even differently if you want) than just overwrite.

And this is common also for non interactive tools such as dpkg (that backups modified configuration files all the times).

So, in case there are situations we didn't consider that the migration may break we still have a safety mechanism.

Reply:
I removed the backup functionality because it caused unexpected behavior when `/etc/groups` is a symlink (it renamed the symlink to `/etc/groups-`, causing `/etc/groups` to be modified instead of `/etc/real-groups`).

I don't see any benefit of restoring a backup file now that we're replacing the group file by writing to a temporary file and then renaming that file. If the rename succeeds, we don't need the backup. If the rename fails, the original group file is preserved, so there is no need for a backup.